### ISSUE TITLE
sql: old sqlalchemy adapter that seems to work with 1.x and 2.0 doesn't with 2.1. The new one does.

### ISSUE THREAD

**Issue by nstewart** [cockroachdb/cockroach#28772]
**BUG REPORT**

Please describe the issue you observed, and any steps we can take to reproduce it:
Sqlalchemy `get_table_names` fails with `Could not locate column in row for column 'Table' in CockroachDB 2.1 alpha` CockroachDB CCL v2.1.0-alpha.20180730

- Which version of CockroachDB are you using?
CockroachDB CCL v2.1.0-alpha.20180730

- What did you do?
Upgraded to the latest cockroachdb alpha and ran the MovR load generator.

```
Traceback (most recent call last):
  File "./loadmovr.py", line 119, in <module>
    exponential_txn_backoff=args.exponential_txn_backoff, echo=args.echo_sql)
  File "/usr/src/app/movr.py", line 22, in __init__
    Base.metadata.drop_all(bind=engine)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/sql/schema.py", line 4031, in drop_all
    tables=tables)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1940, in _run_visitor
    conn._run_visitor(visitorcallable, element, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1549, in _run_visitor
    **kwargs).traverse_single(element)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/sql/visitors.py", line 121, in traverse_single
    return meth(obj, **kw)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/sql/ddl.py", line 844, in visit_metadata
    unsorted_tables = [t for t in tables if self._can_drop_table(t)]
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/sql/ddl.py", line 924, in _can_drop_table
    self.connection, table.name, schema=effective_schema)
  File "/usr/local/lib/python2.7/site-packages/cockroachdb/sqlalchemy/dialect.py", line 95, in has_table
    return any(t == table for t in self.get_table_names(conn, schema=schema))
  File "/usr/local/lib/python2.7/site-packages/cockroachdb/sqlalchemy/dialect.py", line 91, in get_table_names
    return [row.Table for row in conn.execute("SHOW TABLES")]
AttributeError: Could not locate column in row for column 'Table'
```

It looks like in 2.1 we changed the output of `SHOW TABLES` to look like
```
root@localhost:26257/defaultdb> show tables;
+------------+
| table_name |
+------------+
+------------+
```
instead of 
```
root@localhost:26257/movr> show tables;
+-------+
| Table |
+-------+
+-------+
(0 rows)
```

Impact of this is it breaks an existing application.

**Comment by nstewart** [comment]
Yes, looks like I also had to upgrade the `cockroachdb` package. That fixed it!

How can we minimize disruption to production applications during upgrades to CockroachDB 2.1? We definitely need to document this dependency cc @jseldess , but if someone upgrades cockroachdb without upgrading these two packages their apps will just stop working? 

cc @awoods187

**Comment by knz** [comment]
Upgrade sqlalchemy first, then upgrade cockroachdb.

**Comment by knz** [comment]
That's a general requirement btw -- you want the client code to understand both versions before you enter a migration. That's true of any client/server software, not just cockroachdb.

**Comment by nstewart** [comment]
Sure, I get that, but in general, I thought severs weren't supposed to break API contracts when upgrading?

**Comment by jordanlewis** [comment]
The root cause of this is that the SQLAlchemy adapter was programmed to use the `SHOW TABLES` command, which isn't reliable as an API method.

It might not have been explicitly documented, but it's certainly the case that we're not intending to keep the output of any of the `SHOW FOO` commands as fully identical across releases. Clients should use the proper API methods for this - which to a SQL database is the introspection capabilities afforded by `information_schema`, `pg_catalog` and the like. Those are guaranteed to be backward-compatible across releases.

I see two options for us at this point, both of which require fixing the SQLAlchemy adapter and requiring that users update:

1. Revert the change to `SHOW TABLES` for this release, affording users an extra 6 months before the next release to update their SQLAlchemy drivers.
2. Loudly document that SQLAlchemy drivers must be updated before upgrading to CockroachDB 2.1.

Step 2 will be required either way, eventually - in 2.2, if we still want to make this change to `SHOW TABLES`, we'll still need users to ensure they've upgraded their client drivers before upgrading their database.

In conclusion, I don't really think that it's important to revert the change. We're not going to make `SHOW TABLES` immutable for the lifetime of CockroachDB. Let's just make sure there's a callout in the docs about making sure client drivers are upgraded before upgrading CockroachDB itself.

**Comment by knz** [comment]
@nstewart as I also explained to you earlier, until we have people dedicated to maintaining client-side user success (i.e. keeping expertise up to date, running QA, but also understanding how to write client adapters/drivers properly without relying on *expressly unstable server-side APIs* -- that is, not us at this time), we can't offer any guarantee of the kind you're expecting. Things are bound to break.

**Comment by knz** [comment]
In defense of the persons who implemented the sqlalchemy code, back then when this was done, 1) no guarantees were proposed at that time, we actually disclaimed guarantees of backward/forward compatibility and 2) there were not enough server-side APIs to complete a sqlalchemy adapter correctly in the first place 3) we knew there were problems in that implementation already because it was easy to create simple client apps that would fail horribly with that code.

In other words, that piece of sqlalchemy code was released *much* too early. That created wrong expectations (like yours) which we were *never* able to match. 

The 2.0 release would have been a better time to start working on that adapter. We would have done things differently. The new code I put in the sqlalchemy adapter to support 2.0 and 2.1  equally well is more likely to be forward-compatible (we can probably build some confidence on that, although I'm not yet 100% sure). With 2.1 it's getting even better.

**Comment by knz** [comment]
@jseldess @rmloveland I think ultimately we should integrate this in our client/migration docs. Do you want to sit together and make a doc plan about this?

**Comment by jseldess** [comment]
Yes, it'd be great to have your support on this, @knz. As a first step, can you create a corresponding docs issue with a summary of the changes needed? 

For 2.2, documenting [ORMS and retries](https://github.com/cockroachdb/docs/issues/3315) will almost certainly be a docs project (hopefully an eng project, too). This particular update, if I understand right, is more about ensuring users of SQLAlchemy drivers update those before upgrading to CockroachDB v2.1. 

One question I have is in response to this statement by @jordanlewis:

> It might not have been explicitly documented, but it's certainly the case that we're not intending to keep the output of any of the SHOW FOO commands as fully identical across releases. Clients should use the proper API methods for this - which to a SQL database is the introspection capabilities afforded by information_schema, pg_catalog and the like. Those are guaranteed to be backward-compatible across releases.

If the above is true, I think we need to get more explicit in how we document backward-incompatibility. For each case where we do not want to guarantee identical output (or behavior) across releases, we really should state that. In the past, we did that for `SHOW CONSTRAINTS`, which has meanwhile graduate to fully supported. I think SQL features marked `EXPERIMENTAL_` also provide this flexibility.

If we don't mark things with a disclaimer, I think we should feel obligated to treat the change as backward-incompatible and call it out as such in release notes.

CC @bdarnell, who might have thoughts about this, too.

**Comment by knz** [comment]
> If the above is true, I think we need to get more explicit in how we document backward-incompatibility. For each case where we do not want to guarantee identical output (or behavior) across releases, we really should state that

Yes fully agreed.

**Comment by sploiselle** [comment]
@knz Can I get a quick blurb describing this known limitation w/r/t the impact to user experience? Ideally, we need it by Friday 10/26 for the 2.1 Known Limitations page. Posting it on this issue and/or pinging me would be great.

**Comment by awoods187** [comment]
Can we make sure this works for 19.1 and that SQLAlchemy support will be ready for our users? I think this ties in well with recent work @BramGruneir has been doing with the test suite

**Comment by jordanlewis** [comment]
This has been resolved for 2.1 and 19.1 at this point.

**Comment by jseldess** [comment]
@jordanlewis, does your comment above mean we can remove this limitation from the 19.2 and 20.1 docs? https://www.cockroachlabs.com/docs/stable/known-limitations.html#using-sqlalchemy-with-cockroachdb

**Comment by jordanlewis** [comment]
I think so but I'll defer to @rafiss.

**Comment by rafiss** [comment]
The fix for this issue was an update to our sqlalchemy-cockroachdb adapter. So we should still tell people to use version 0.2.1 of the adapter or later, even for 19.2 and 20.1.

The latest version of the adapter (currently 0.4.0) will have all the fixes for other issues, so we should still advise people to use that, but only >=0.2.1 is strictly required.

**Comment by jseldess** [comment]
Thanks, Rafi. I think I'll leave the limitation generic and say that you should upgrade to the latest version before upgrading your cluster.
